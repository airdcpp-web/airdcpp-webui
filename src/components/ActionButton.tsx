import Button, { ButtonProps } from 'components/semantic/Button';
import { showAction, toActionI18nKey } from 'utils/ActionUtils';
import { IconType } from 'components/semantic/Icon';

import * as UI from 'types/ui';
import { ActionHandlerDecorator } from 'decorators/ActionHandlerDecorator';
import { useTranslation } from 'react-i18next';

export interface ActionButtonProps<ItemDataT extends UI.ActionItemDataValueType>
  extends Omit<ButtonProps, 'caption' | 'icon'> {
  actions: UI.ModuleActions<ItemDataT>;
  actionId: string;
  itemData?: ItemDataT;
  icon?: IconType | true;
}

const ActionButton = <ItemDataT extends UI.ActionItemDataValueType>({
  actionId,
  itemData,
  icon = true,
  actions,
  ...other
}: ActionButtonProps<ItemDataT>) => {
  const action = actions.actions[actionId]!;
  const { t } = useTranslation();
  if (!showAction(action, itemData)) {
    return null;
  }

  const { moduleId, subId } = actions;
  return (
    <ActionHandlerDecorator<ItemDataT>>
      {({ onClickAction }) => (
        <Button
          icon={icon ? (typeof icon === 'string' ? icon : action.icon) : null}
          onClick={() =>
            onClickAction({
              itemData: itemData as ItemDataT,
              action,
              actionId,
              moduleId,
              subId,
            })
          }
          caption={t(toActionI18nKey(action, moduleId), action.displayName)}
          {...other}
        />
      )}
    </ActionHandlerDecorator>
  );
};

/*ActionButton.propTypes = {
  action: PropTypes.func,

  itemData: PropTypes.object,

  icon: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.string,
  ]),
};*/

//export default ActionHandlerDecorator<ActionButtonProps<ItemDataT>, ItemDataT>(ActionButton);
export default ActionButton;
