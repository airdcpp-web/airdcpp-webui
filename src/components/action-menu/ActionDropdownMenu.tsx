import * as React from 'react';

import SectionedDropdown, {
  SectionedDropdownProps,
} from '@/components/semantic/SectionedDropdown';
import MenuSection from '@/components/semantic/MenuSection';
import ActionMenuDecorator, {
  ActionMenuDecoratorChildProps,
} from '@/components/action-menu/decorators/ActionMenuDecorator';
import { buildMenu } from './builder/expandMenuBuilder';

export interface ActionMenuProps extends SectionedDropdownProps {
  header?: React.ReactNode;
}

const ActionDropdownMenu = ({
  header,
  children,
  ...other
}: ActionMenuProps & ActionMenuDecoratorChildProps) => (
  <SectionedDropdown {...other}>
    <MenuSection caption={header}>{children()}</MenuSection>
  </SectionedDropdown>
);

const ActionMenuDecorated = ActionMenuDecorator<ActionMenuProps, any, any>(
  ActionDropdownMenu,
  buildMenu,
);

export { ActionMenuDecorated as ActionMenu };
