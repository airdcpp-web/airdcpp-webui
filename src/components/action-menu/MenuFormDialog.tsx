import { useRef } from 'react';
import * as React from 'react';

import Modal, { ModalProps } from '@/components/semantic/Modal';
import Form, { FormSaveHandler } from '@/components/form/Form';

import * as UI from '@/types/ui';

import { IconType } from '@/components/semantic/Icon';
import { useTranslation } from 'react-i18next';
import { translate } from '@/utils/TranslationUtils';

export interface MenuFormDialogProps extends ModalProps {
  title: string;
  icon: IconType;
  fieldDefinitions: UI.FormFieldDefinition[];
  onSave: FormSaveHandler<UI.FormValueMap>;
}

export const MenuFormDialog: React.FC<MenuFormDialogProps> = ({
  fieldDefinitions,
  onSave,
  title,
  icon,
  ...other
}) => {
  const formRef = useRef<Form | null>(null);
  const { t } = useTranslation();
  return (
    <Modal
      {...other}
      title={title}
      onApprove={() => formRef.current!.save()}
      approveCaption={translate('Continue', t, UI.Modules.COMMON)}
      closable={false}
      icon={icon}
      dynamicHeight={true}
    >
      <Form ref={formRef} onSave={onSave} fieldDefinitions={fieldDefinitions} />
    </Modal>
  );
};
