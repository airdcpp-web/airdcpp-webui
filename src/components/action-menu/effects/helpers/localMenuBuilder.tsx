import { toActionI18nKey } from '@/utils/ActionUtils';

import * as UI from '@/types/ui';

import { ActionClickHandler } from '@/decorators/ActionHandlerDecorator';
import { Trans } from 'react-i18next';
import { parseTranslationModules } from '@/utils/TranslationUtils';

const getDivider = (menuIndex: number, itemIndex: number) => {
  return {
    id: `divider${menuIndex}_${itemIndex}`,
  };
};

const actionToActionMenuItem = <
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
>(
  action: UI.MenuActionDefition<ItemDataT, EntityT>,
  menu: UI.ActionMenuType<ItemDataT, EntityT>,
  onClickAction: ActionClickHandler<ItemDataT, EntityT>,
) => {
  const active = !action.checked ? false : action.checked(menu.itemDataGetter());
  const icon = !!action.checked ? (active ? 'checkmark' : '') : action.icon;
  return {
    id: action.id,
    item: {
      onClick: () => {
        onClickAction({
          action,
          itemData: menu.itemDataGetter(),
          entity: menu.entity,
          moduleData: menu.moduleData,
        });
      },
      active,
      icon,
      children: (
        <Trans
          i18nKey={toActionI18nKey(
            action,
            parseTranslationModules(menu.moduleData.moduleId),
          )}
          defaults={action.displayName}
        >
          {action.displayName}
        </Trans>
      ),
    },
  };
};

// Convert ID to menu link element
const getMenuItem = <
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
>(
  menu: UI.ActionMenuType<ItemDataT, EntityT>,
  menuIndex: number,
  action: UI.MenuActionListItemType<ItemDataT, EntityT>,
  itemIndex: number,
  onClickAction: ActionClickHandler<ItemDataT, EntityT>,
): UI.ActionMenuItem => {
  if (!action) {
    return getDivider(menuIndex, itemIndex);
  }

  if ('children' in action) {
    const children = action.children.map((childAction, childActionIndex) => {
      if (!childAction) {
        return getDivider(0, childActionIndex);
      }

      return actionToActionMenuItem(childAction, menu, onClickAction);
    });

    if (children.length === 1) {
      return children[0];
    }

    return {
      id: action.id,
      item: {
        icon: action.icon,
        onClick: () => {
          // ..
        },
        children: <>{action.displayName}</>,
      },
      children,
    };
  }

  return actionToActionMenuItem(action, menu, onClickAction);
};

export const localMenuToActionMenuItems = <
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
>(
  onClickAction: ActionClickHandler<ItemDataT, EntityT>,
  menu: UI.ActionMenuType<ItemDataT, EntityT>,
  menuIndex: number,
) => {
  const items = menu.actions.map((actionId, actionIndex) => {
    return getMenuItem(
      menu as UI.ActionMenuType<ItemDataT, EntityT>,
      menuIndex,
      actionId,
      actionIndex,
      onClickAction,
    );
  });

  return items;
};
