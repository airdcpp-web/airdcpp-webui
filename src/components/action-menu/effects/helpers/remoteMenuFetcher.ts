import { useEffect, useState } from 'react';

import * as API from 'types/api';
import * as UI from 'types/ui';

import MenuConstants, { MENU_SUPPORTS } from 'constants/MenuConstants';
import NotificationActions from 'actions/NotificationActions';
import { parseActionMenuItemId, parseActionMenuItemIds } from 'utils/MenuUtils';
import { ActionMenuDefinition } from '../useActionMenuItems';
import { useSocket } from 'context/SocketContext';

export interface RemoteMenuData {
  selectedIds: UI.ActionIdType[];
  remoteMenuId: string;
  entityId: API.IdType | undefined;
}

export interface RemoteMenu {
  menuData: RemoteMenuData;
  items: API.GroupedContextMenuItem[];
}

interface ParsedDefinitionArray {
  remoteMenuIds: Array<string | undefined>;
  selectedIdsByMenu: Array<UI.ActionIdType>[];
  entityIds: Array<API.IdType | undefined>;
}

const parseDefinitionArray = (
  definitionArray: ActionMenuDefinition<any, any>[],
): ParsedDefinitionArray => {
  const selectedIdsByMenu = definitionArray
    .map((defition) => defition.itemData)
    .map((data) => parseActionMenuItemIds(data));
  const remoteMenuIds = definitionArray.map(
    (remoteMenuProps) => remoteMenuProps.remoteMenuId,
  );
  const entityIds = definitionArray
    .map((defition) => defition.entity)
    .map((data) => parseActionMenuItemId(data) as API.IdType | undefined);

  return { selectedIdsByMenu, remoteMenuIds, entityIds };
};

export const useRemoteMenuFetcher = (
  definitionArray: ActionMenuDefinition<any, any>[],
) => {
  const { selectedIdsByMenu, remoteMenuIds, entityIds } =
    parseDefinitionArray(definitionArray);

  const [menusByType, setMenusByType] = useState<
    Array<RemoteMenu | undefined> | undefined
  >(undefined);

  const socket = useSocket();

  useEffect(() => {
    const fetchMenus = async () => {
      const menuPromises = remoteMenuIds.map(async (remoteMenuId, menuIndex) => {
        if (!remoteMenuId) {
          return Promise.resolve(undefined);
        }

        const selectedIds = selectedIdsByMenu[menuIndex];
        const entityId = entityIds[menuIndex];

        const menu = await socket.post<API.GroupedContextMenuItem[]>(
          `${MenuConstants.MODULE_URL}/${remoteMenuId}/list_grouped`,
          {
            selected_ids: selectedIds,
            entity_id: entityId,
            supports: MENU_SUPPORTS,
          },
        );

        return {
          items: menu,
          menuData: {
            remoteMenuId,
            selectedIds,
            entityId,
          },
        };
      });

      let fetchedMenus: Array<RemoteMenu | undefined>;
      try {
        fetchedMenus = await Promise.all(menuPromises);
      } catch (e) {
        NotificationActions.apiError('Failed fetch menu items', e);
        return;
      }

      setMenusByType(fetchedMenus);
    };

    fetchMenus();
  }, []);

  return menusByType;
};
