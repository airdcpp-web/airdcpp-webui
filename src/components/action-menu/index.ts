export * from './ActionDropdownMenu';
export * from './DropdownMenu';
export * from '../icon/MenuIcon';
