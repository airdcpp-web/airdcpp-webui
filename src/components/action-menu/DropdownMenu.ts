import TableDropdown, { TableDropdownProps } from 'components/semantic/TableDropdown';

import ActionMenuDecorator, {
  ActionMenuDecoratorProps,
} from './decorators/ActionMenuDecorator';
import DownloadMenuDecorator, {
  DownloadMenuDecoratorProps,
} from './decorators/DownloadMenuDecorator';
import UserMenuDecorator, {
  UserMenuDecoratorProps,
} from './decorators/UserMenuDecorator';

import { DropdownProps } from 'components/semantic/Dropdown';

import * as UI from 'types/ui';
import { ActionMenuProps, ActionMenu } from './ActionDropdownMenu';
import { buildMenu } from './builder/slidingMenuBuilder';

type TableActionMenuDropdownProps = Omit<TableDropdownProps, 'children' | 'caption'>;

export type TableActionMenuProps<ItemDataT extends UI.ActionMenuItemDataValueType> =
  TableActionMenuDropdownProps & ActionMenuDecoratorProps<ItemDataT>;
export const TableActionMenu = ActionMenuDecorator<TableActionMenuDropdownProps, any>(
  TableDropdown,
  buildMenu,
);

export type UserMenuProps = UserMenuDecoratorProps & ActionMenuProps;
export const UserMenu = UserMenuDecorator<DropdownProps>(ActionMenu);

export type TableUserMenuProps = UserMenuDecoratorProps & TableActionMenuDropdownProps;
export const TableUserMenu =
  UserMenuDecorator<TableActionMenuDropdownProps>(TableActionMenu);

export type DownloadMenuProps<ItemDataT extends UI.DownloadableItemInfo> =
  DownloadMenuDecoratorProps<ItemDataT> & ActionMenuProps;

export const DownloadMenu = DownloadMenuDecorator<DropdownProps, any>(ActionMenu);

export type TableDownloadMenuProps<ItemDataT extends UI.DownloadableItemInfo> =
  DownloadMenuDecoratorProps<ItemDataT> & TableActionMenuDropdownProps;

export const TableDownloadMenu = DownloadMenuDecorator<TableActionMenuDropdownProps, any>(
  TableActionMenu,
);
