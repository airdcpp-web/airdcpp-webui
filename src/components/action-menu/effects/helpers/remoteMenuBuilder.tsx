import * as API from '@/types/api';
import * as UI from '@/types/ui';

import MenuConstants from '@/constants/MenuConstants';
import IconConstants from '@/constants/IconConstants';
import { RemoteMenu, RemoteMenuData } from './remoteMenuFetcher';
import { UILogger } from '@/effects/LoggerEffect';

export interface CombinedContextMenuItem {
  typeData: RemoteMenuData;
  item: API.ContextMenuItem;
}

export interface CombinedContextMenu extends Omit<API.GroupedContextMenu, 'items'> {
  items: CombinedContextMenuItem[];
}

interface MenuOptions {
  handleSelect: SelectHandler;
  nestingThreshold: number;
}

type SelectHandler = (
  menuItem: API.ContextMenuItem,
  remoteMenuData: RemoteMenuData,
) => void;

const toMenuItem = (
  menuItem: API.ContextMenuItem,
  remoteMenuData: RemoteMenuData,
  handleSelect: SelectHandler,
): UI.ActionMenuItem => {
  const onClick = () => {
    handleSelect(menuItem, remoteMenuData);
  };

  const { id, icon, title } = menuItem;
  return {
    id,
    item: {
      onClick,
      icon: icon[MenuConstants.ICON_TYPE_ID],
      children: title,
    },
  };
};

export const remoteMenuToActionMenuItems = (
  menu: CombinedContextMenu,
  { handleSelect, nestingThreshold }: MenuOptions,
): UI.ActionMenuItem[] => {
  const items = menu.items.map(({ item, typeData }) =>
    toMenuItem(item, typeData, handleSelect),
  );

  if (items.length <= nestingThreshold) {
    return items;
  }

  return [
    {
      id: menu.id,
      item: {
        icon: menu.icon[MenuConstants.ICON_TYPE_ID] || IconConstants.EXTENSION,
        onClick: () => {
          // ..
        },
        children: <>{menu.title}</>,
      },
      children: items,
    },
  ];
};

// Combine menus by their subscriber ID and remove duplicate items
export const createRemoteMenuCombineReducer = (logger: UILogger) => {
  return (reduced: CombinedContextMenu[], typeMenus: RemoteMenu | undefined) => {
    if (!typeMenus) {
      return reduced;
    }

    for (const menu of typeMenus.menus) {
      const existing = reduced.find((m) => m.id === menu.id);

      const toMenuItem = (item: API.ContextMenuItem): CombinedContextMenuItem => ({
        typeData: typeMenus.typeData,
        item,
      });

      if (existing) {
        existing.items.push(
          ...menu.items
            .filter((item) => {
              const existingItem = existing.items.find(
                (existingItem) =>
                  existingItem.item.id === item.id &&
                  existingItem.item.hook_id === item.hook_id,
              );
              if (existingItem) {
                logger.info(
                  // eslint-disable-next-line max-len
                  `Removing a duplicate item ${item.id} (subscriber ${menu.id}) from menu ${typeMenus.typeData.remoteMenuId} (exists for menu ${existingItem.typeData.remoteMenuId})`,
                );
              }

              return !existingItem;
            })
            .map(toMenuItem),
        );
      } else {
        reduced.push({
          ...menu,
          items: menu.items.map(toMenuItem),
        });
      }
    }

    return reduced;
  };
};
