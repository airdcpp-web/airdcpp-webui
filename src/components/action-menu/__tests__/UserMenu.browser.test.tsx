import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { renderDataRoutes } from '@/tests/render/test-renderers';

import { waitFor } from '@testing-library/dom';

import {
  clickMenuItem,
  expectRequestToMatchSnapshot,
} from '@/tests/helpers/test-helpers';
import { setupUserEvent } from '@/tests/helpers/test-form-helpers';
import { initCommonDataMocks } from '@/tests/mocks/mock-data-common';
import { getMockServer, MockServer } from '@/tests/mocks/mock-server';
import {
  RemoteMenu2Item,
  RemoteMenuGrouped1,
  RemoteMenuGrouped2,
} from '@/tests/mocks/api/menu';
import { ActionMenu, UserMenu } from '@/components/action-menu';
import { VIEW_FIXED_HEIGHT } from '@/tests/render/test-containers';
import { MockUser1Response } from '@/tests/mocks/api/user';
import UserConstants from '@/constants/UserConstants';
import { getTestActionMenu, installActionMenuMocks } from './test-action-menu-helpers';

const RemoteMenuId1 = 'test-remote-menu-id';
const RemoteMenuId2 = 'test-remote-menu-id-2';

// tslint:disable:no-empty
describe('User menu', () => {
  let server: MockServer;

  beforeEach(() => {
    server = getMockServer();
  });

  afterEach(() => {
    server.stop();
  });

  const getSocket = async () => {
    const commonData = await initCommonDataMocks(server);

    // User handlers
    const onIgnoreUser = vi.fn();
    server.addRequestHandler(
      'POST',
      `${UserConstants.IGNORES_URL}/${MockUser1Response.cid}`,
      undefined,
      onIgnoreUser,
    );

    // Remote menu mocks
    const menu1Mocks = installActionMenuMocks(
      RemoteMenuId1,
      [RemoteMenuGrouped1],
      server,
    );

    const userEvent = setupUserEvent();
    return {
      commonData,
      server,

      menu1Mocks,

      onIgnoreUser,
      userEvent,
    };
  };

  const renderMenu = async (enabledRemoteMenu2?: boolean) => {
    const { commonData, server, ...other } = await getSocket();

    const { menu, ...actionHandlers } = getTestActionMenu();

    const onSave = vi.fn(() => Promise.resolve());

    const UserDropdownMenuTest = () => {
      const testItemData = { id: 'test', name: 'Test Item' };
      const testEntity = { id: 'test-entity', name: 'Test Entity' };
      return (
        <UserMenu
          remoteMenuId={RemoteMenuId1}
          user={MockUser1Response}
          label={MockUser1Response.nicks}
        >
          <ActionMenu
            actions={menu}
            itemData={testItemData}
            entity={testEntity}
            remoteMenuId={enabledRemoteMenu2 ? RemoteMenuId2 : undefined}
            remoteMenuNestingThreshold={2}
          />
        </UserMenu>
      );
    };

    const routes = [
      {
        index: true,
        Component: UserDropdownMenuTest,
      },
    ];

    const renderData = renderDataRoutes(routes, commonData, {
      routerProps: { initialEntries: ['/'] },
      viewType: VIEW_FIXED_HEIGHT,
    });

    return { onSave, ...renderData, ...other, ...actionHandlers };
  };

  const clickItem = async (
    menuItemLabel: string,
    renderResult: Awaited<ReturnType<typeof renderMenu>>,
  ) => {
    const { getByRole } = renderResult;
    await waitFor(() =>
      expect(getByRole('menuitem', { name: menuItemLabel })).toBeTruthy(),
    );

    // Click
    await clickMenuItem(menuItemLabel, renderResult);
  };

  const openMenu = async (renderResult: Awaited<ReturnType<typeof renderMenu>>) => {
    const { getByLabelText, getByRole, userEvent } = renderResult;
    await waitFor(() => expect(getByLabelText(MockUser1Response.nicks)).toBeTruthy());

    await userEvent.click(getByLabelText(MockUser1Response.nicks));

    await waitFor(() => expect(getByRole('menu')).toBeTruthy());
  };

  const waitMenuClosed = async ({
    queryByRole,
  }: Awaited<ReturnType<typeof renderMenu>>) => {
    await waitFor(() => expect(queryByRole('menu')).not.toBeInTheDocument());
  };

  describe('Standalone', () => {
    test('should select local item', async () => {
      const renderResult = await renderMenu();
      const { onIgnoreUser } = renderResult;

      await openMenu(renderResult);
      await clickItem('Ignore messages', renderResult);

      await waitFor(() => expect(onIgnoreUser).toHaveBeenCalledTimes(1));
      await waitMenuClosed(renderResult);
    }, 100000);
  });

  describe('Nested', () => {
    test('should select menu item', async () => {
      const menu2Mocks = installActionMenuMocks(
        RemoteMenuId2,
        [RemoteMenuGrouped2],
        server,
      );

      const renderResult = await renderMenu(true);

      await openMenu(renderResult);

      await waitFor(() => expect(menu2Mocks.onListGrouped).toHaveBeenCalledTimes(1));

      await clickItem(RemoteMenuGrouped1.title, renderResult);

      await clickItem(RemoteMenu2Item.title, renderResult);
      await waitFor(() => expect(menu2Mocks.onSelectMenuItem).toHaveBeenCalledTimes(1));

      expectRequestToMatchSnapshot(menu2Mocks.onSelectMenuItem);
      await waitMenuClosed(renderResult);
    }, 100000);

    test('should merge duplicate menus', async () => {
      const renderResult = await renderMenu(true);

      const menu2Mocks = installActionMenuMocks(
        RemoteMenuId2,
        [RemoteMenuGrouped1, RemoteMenuGrouped2],
        server,
      );

      const { findAllByText, menu1Mocks } = renderResult;

      await openMenu(renderResult);

      await waitFor(() => expect(menu1Mocks.onListGrouped).toHaveBeenCalledTimes(1));
      await waitFor(() => expect(menu2Mocks.onListGrouped).toHaveBeenCalledTimes(1));

      const menu1Submenus = await findAllByText(RemoteMenuGrouped1.title);
      expect(menu1Submenus).toHaveLength(1);

      const menu2Items = await findAllByText(RemoteMenu2Item.title);
      expect(menu2Items).toHaveLength(1);

      await clickItem(RemoteMenuGrouped1.title, renderResult);

      await clickItem(RemoteMenu2Item.title, renderResult);
      await waitFor(() => expect(menu2Mocks.onSelectMenuItem).toHaveBeenCalledTimes(1));

      expectRequestToMatchSnapshot(menu2Mocks.onSelectMenuItem);
      await waitMenuClosed(renderResult);
    }, 100000);
  });
});
