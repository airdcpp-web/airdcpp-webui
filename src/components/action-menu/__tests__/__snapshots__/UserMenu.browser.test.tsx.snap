// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`User menu > Nested > should merge duplicate menus 1`] = `
{
  "data": {
    "entity_id": "test-entity",
    "form_definitions": null,
    "hook_id": "airdcpp-release-validator",
    "menuitem_id": "scan_missing_extra",
    "selected_ids": [
      "test",
    ],
    "supports": [
      "urls",
      "form",
    ],
  },
  "method": "POST",
  "path": "menus/test-remote-menu-id-2/select",
}
`;

exports[`User menu > Nested > should select menu item 1`] = `
{
  "data": {
    "entity_id": "test-entity",
    "form_definitions": null,
    "hook_id": "airdcpp-release-validator",
    "menuitem_id": "scan_missing_extra",
    "selected_ids": [
      "test",
    ],
    "supports": [
      "urls",
      "form",
    ],
  },
  "method": "POST",
  "path": "menus/test-remote-menu-id-2/select",
}
`;
